src/java/org/apache/cassandra/db/ColumnFamilyStore.java

```diff
@@ -453,8 +454,12 @@ public void updateSpeculationThreshold()
     {
         try
         {
-            sampleReadLatencyNanos = metadata().params.speculativeRetry.calculateThreshold(metric.coordinatorReadLatency);
-            transientWriteLatencyNanos = metadata().params.speculativeWriteThreshold.calculateThreshold(metric.coordinatorWriteLatency);
+            Snapshot readLatencySnapshot = metric.coordinatorReadLatency.getSnapshot();
+            if (readLatencySnapshot.size() > 0)
+                sampleReadLatencyNanos = metadata().params.speculativeRetry.calculateThreshold(readLatencySnapshot);
+            Snapshot writeLatencySnapshot = metric.coordinatorWriteLatency.getSnapshot();
+            if (writeLatencySnapshot.size() > 0)
+                transientWriteLatencyNanos = metadata().params.speculativeWriteThreshold.calculateThreshold(writeLatencySnapshot);
```

Comment:
Since the size > 0 is really a precondition for calculateThreshold to run successfully, it seems like the check should be there instead in the policies that require it vs an implicit undocumented precondition to be checked before in only the existing calls (incase a new entry point ever created).

Reply:
You are right for fixed policies it does the wrong thing. Good catch.